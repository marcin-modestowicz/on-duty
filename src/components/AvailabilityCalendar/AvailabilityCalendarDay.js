// @flow
import React, { Component } from "react";
import { observer } from "mobx-react";
import classnames from "classnames";
import type Availability, {
  AvailabilityStatus
} from "../../models/Availability";
import {
  AVAILABILITY_STATUSES,
  AVAILABILITY_STATUSES_NAMES
} from "../../models/Availability";
import styles from "./AvailabilityCalendarDay.scss";

type Props = {
  day: number,
  availability: Availability
};

@observer
class AvailabilityCalendarDay extends Component<Props> {
  statuses: AvailabilityStatus[] = [
    AVAILABILITY_STATUSES.FREE,
    AVAILABILITY_STATUSES.KEEN,
    AVAILABILITY_STATUSES.BUSY
  ];

  handleClick = () => {
    this.props.availability.toggle();
  };

  renderDayContent(status: string) {
    return (
      <div key={status} className={classnames(styles.day, styles[status])}>
        <div className={styles.dayName}>Monday</div>
        <div className={styles.dayNum}>{this.props.day}</div>
      </div>
    );
  }

  render() {
    const status = this.props.availability.status;
    const isFreeActive = status === AVAILABILITY_STATUSES.FREE;
    const isKeenActive = status === AVAILABILITY_STATUSES.KEEN;
    const isBusyActive = status === AVAILABILITY_STATUSES.BUSY;

    return (
      <div
        className={classnames(styles.root, {
          [styles.freeActive]: isFreeActive,
          [styles.keenActive]: isKeenActive,
          [styles.busyActive]: isBusyActive
        })}
        role="button"
        title="Click to toggle availability"
        onClick={this.handleClick}
      >
        {this.statuses.map(_status =>
          this.renderDayContent(
            AVAILABILITY_STATUSES_NAMES[_status].toLowerCase()
          )
        )}
      </div>
    );
  }
}

export default AvailabilityCalendarDay;
