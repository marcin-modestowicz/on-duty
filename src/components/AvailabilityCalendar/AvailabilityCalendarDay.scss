@import './colors.scss';

.root {
  position: relative;
  width: 5em;
  height: 5em;
  margin: .25em;

  .day {
    box-sizing: border-box;
    position: absolute;
    border-width: 0 1px 1px 0;
    border-style: solid;
    border-radius: .25em;
    width: 5em;
    height: 5em;
    padding: .75em;
    cursor: pointer;
    user-select: none;
    text-align: center;
    color: white;
    transition: 0.6s;
    transform-style: preserve-3d;
    backface-visibility: hidden;

    .name {
      margin-bottom: .5em;
      font-size: .5em;
    }

    .number {
      font-size: 2em;
      font-weight: bold;

      &:after {
        content: '';
        display: block;
        margin-top: .1em;
        width: 100%;
        height: 1px;
        border-bottom: 1px solid;
      }
    }

    &.free {
      background-color: $free;
      border-color: $free-darker;

      .dayNum:after {
        background-color: $free-darker;
        border-color: $free-lighter;
      }
    }

    &.keen {
      background-color: $keen;
      border-color: $keen-darker;

      .dayNum:after {
        background-color: $keen-darker;
        border-color: $keen-lighter;
      }
    }

    &.busy {
      background-color: $busy;
      border-color: $busy-darker;

      .dayNum:after {
        background-color: $busy-darker;
        border-color: $busy-lighter;
      }
    }
  }

  &.freeActive {
    .free {
      transform: rotateY(0deg);
    }

    .keen {
      transform: rotateY(-180deg);
    }

    .busy {
      transform: rotateY(-180deg);
    }
  }

  &.keenActive {
    .free {
      transform: rotateY(-180deg);
    }

    .keen {
      transform: rotateY(0deg);
    }

    .busy {
      transform: rotateY(-180deg);
    }
  }

  &.busyActive {
    .free {
      transform: rotateY(-180deg);
    }

    .keen {
      transform: rotateY(-180deg);
    }

    .busy {
      transform: rotateY(0deg);
    }
  }
}
