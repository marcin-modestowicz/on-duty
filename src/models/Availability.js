// @flow
import { observable, computed, action } from "mobx";

export const AVAILABILITY_STATUSES = {
  FREE: 0,
  KEEN: 1,
  BUSY: -1
};

export const AVAILABILITY_STATUSES_NAMES = {
  "0": "Free",
  "1": "Keen",
  "-1": "Busy"
};

export type AvailabilityStatus = -1 | 0 | 1;

class Availability {
  @observable status: AvailabilityStatus;

  constructor(availability?: AvailabilityStatus) {
    this.status = availability || AVAILABILITY_STATUSES.FREE;
  }

  @computed
  get statusName(): AvailabilityStatus {
    return AVAILABILITY_STATUSES_NAMES[this.status];
  }

  @action
  toggle() {
    switch (this.status) {
      case AVAILABILITY_STATUSES.BUSY:
        this.status = AVAILABILITY_STATUSES.FREE;
        break;
      case AVAILABILITY_STATUSES.KEEN:
        this.status = AVAILABILITY_STATUSES.BUSY;
        break;
      default:
        this.status = AVAILABILITY_STATUSES.KEEN;
    }
  }

  @action
  setStatus(status: AvailabilityStatus) {
    this.status = status;
  }
}

export default Availability;
