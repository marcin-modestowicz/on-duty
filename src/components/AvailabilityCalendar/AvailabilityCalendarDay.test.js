// @flow
import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Availability from "../../models/Availability";
import AvailabilityCalendarDay from "./AvailabilityCalendarDay";

describe("AvailabilityCalendarDay component", () => {
  let availability;
  let availabilityCalendarDay;

  beforeAll(() => {
    availability = new Availability();
    availabilityCalendarDay = shallow(
      <AvailabilityCalendarDay
        date={new Date(2017, 9, 1)}
        availability={availability}
      />
    );
  });

  test("should render with free status active", () => {
    expect(toJson(availabilityCalendarDay)).toMatchSnapshot();
  });

  test("should render with keen status active", () => {
    availability.toggle();
    expect(toJson(availabilityCalendarDay)).toMatchSnapshot();
  });

  test("should render with busy status active", () => {
    availability.toggle();
    expect(toJson(availabilityCalendarDay)).toMatchSnapshot();
  });
});
