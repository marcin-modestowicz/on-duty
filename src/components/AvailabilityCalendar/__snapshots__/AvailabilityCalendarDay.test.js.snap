// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvailabilityCalendarDay component should render with busy status active 1`] = `
<div
  className="root busyActive"
  onClick={[Function]}
  role="button"
  title="Click to toggle availability"
>
  <div
    className="day free"
    key="free"
  >
    <div
      className="name"
    >
      Sunday
    </div>
    <div
      className="number"
    >
      1
    </div>
  </div>
  <div
    className="day keen"
    key="keen"
  >
    <div
      className="name"
    >
      Sunday
    </div>
    <div
      className="number"
    >
      1
    </div>
  </div>
  <div
    className="day busy"
    key="busy"
  >
    <div
      className="name"
    >
      Sunday
    </div>
    <div
      className="number"
    >
      1
    </div>
  </div>
</div>
`;

exports[`AvailabilityCalendarDay component should render with free status active 1`] = `
<div
  className="root freeActive"
  onClick={[Function]}
  role="button"
  title="Click to toggle availability"
>
  <div
    className="day free"
    key="free"
  >
    <div
      className="name"
    >
      Sunday
    </div>
    <div
      className="number"
    >
      1
    </div>
  </div>
  <div
    className="day keen"
    key="keen"
  >
    <div
      className="name"
    >
      Sunday
    </div>
    <div
      className="number"
    >
      1
    </div>
  </div>
  <div
    className="day busy"
    key="busy"
  >
    <div
      className="name"
    >
      Sunday
    </div>
    <div
      className="number"
    >
      1
    </div>
  </div>
</div>
`;

exports[`AvailabilityCalendarDay component should render with keen status active 1`] = `
<div
  className="root keenActive"
  onClick={[Function]}
  role="button"
  title="Click to toggle availability"
>
  <div
    className="day free"
    key="free"
  >
    <div
      className="name"
    >
      Sunday
    </div>
    <div
      className="number"
    >
      1
    </div>
  </div>
  <div
    className="day keen"
    key="keen"
  >
    <div
      className="name"
    >
      Sunday
    </div>
    <div
      className="number"
    >
      1
    </div>
  </div>
  <div
    className="day busy"
    key="busy"
  >
    <div
      className="name"
    >
      Sunday
    </div>
    <div
      className="number"
    >
      1
    </div>
  </div>
</div>
`;
